File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        left = 0
        counter = [0]*26
        maxLen = 0
        maxCount = 0
        for right in range(0, len(s)):
            counter[ord(s[right]) - 65] += 1
            maxCount = max(maxCount, counter[ord(s[right]) - 65])
            while right - left + 1 - maxCount > k:
                counter[ord(s[left]) - 65] -= 1
                left += 1
            maxLen = max(maxLen, right - left + 1)
        return maxLen

File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int n = s.length();
        HashMap<Character, Integer> count = new HashMap<>();
        int left = 0;
        int maxUniqueCount = 0;
        int ans = 0;
        for (int right = 0; right < n; right++) {
            Character c = s.charAt(right);
            count.put(c, count.getOrDefault(c, 0) + 1);
            maxUniqueCount = Math.max(maxUniqueCount, count.get(c));
            while (right - left + 1 > maxUniqueCount + k) {
                count.put(s.charAt(left), count.get(s.charAt(left)) - 1);
                left++;
                for (Character key : count.keySet()) {
                    maxUniqueCount = Math.max(maxUniqueCount, count.get(key));
                }
            }
            ans = Math.max(ans, right - left + 1);
        }
        return ans;
    }
}